plugin/cache/prefech_test.go

```diff
@@ -2,53 +2,181 @@ package cache
+				},
+			},
+		},
+		{
+			qname:    "hits.reset.",
+			ttl:      80,
+			prefetch: 1,
+			verifications: []verification{
+				{
+					after:  0 * time.Second,
+					answer: "hits.reset. 80 IN A 127.0.0.1",
+					fetch:  true,
+				},
+				{
+					after:  73 * time.Second,
+					answer: "hits.reset. 7 IN A 127.0.0.1",
+					fetch:  true,
+				},
+				{
+					after:  80 * time.Second,
+					answer: "hits.reset. 73 IN A 127.0.0.2",
+				},
+			},
+		},
+		{
+			qname:    "redis.service.dc1.consul.",
+			ttl:      5,
+			prefetch: 1,
+			verifications: []verification{
+				{
+					after:  0 * time.Second,
+					answer: "redis.service.dc1.consul. 5 IN A 127.0.0.1",
+					fetch:  true,
+				},
+				{
+					after:  1 * time.Second,
+					answer: "redis.service.dc1.consul. 4 IN A 127.0.0.1",
+				},
+				{
+					after:  4 * time.Second,
+					answer: "redis.service.dc1.consul. 1 IN A 127.0.0.1",
+					fetch:  true,
+				},
+				{
+					after:  5 * time.Second,
+					answer: "redis.service.dc1.consul. 4 IN A 127.0.0.2",
+				},
+			},
+		},
+		{
+			qname:    "no.prefetch.",
+			ttl:      30,
+			prefetch: 0,
+			verifications: []verification{
+				{
+					after:  0 * time.Second,
+					answer: "no.prefetch. 30 IN A 127.0.0.1",
+					fetch:  true,
+				},
+				{
+					after:  15 * time.Second,
+					answer: "no.prefetch. 15 IN A 127.0.0.1",
+				},
+				{
+					after:  29 * time.Second,
+					answer: "no.prefetch. 1 IN A 127.0.0.1",
+				},
+				{
+					after:  30 * time.Second,
+					answer: "no.prefetch. 30 IN A 127.0.0.2",
+					fetch:  true,
+				},
+			},
+		},
+	}
 
-	ctx := context.TODO()
+	t0, err := time.Parse(time.RFC3339, "2018-01-01T14:00:00+00:00")
+	if err != nil {
+		t.Fatal(err)
+	}
+	fetchc := make(chan struct{}, 1)
+	for _, tt := range tests {
+		c := NewCache()
+		c.prefetch = tt.prefetch
+		c.Next = prefetchHandler(tt.qname, tt.ttl, fetchc)
 
-	req := new(dns.Msg)
-	req.SetQuestion("lowttl.example.org.", dns.TypeA)
+		req := new(dns.Msg)
+		req.SetQuestion(tt.qname, dns.TypeA)
+		rec := dnstest.NewRecorder(&test.ResponseWriter{})
 
-	rec := dnstest.NewRecorder(&test.ResponseWriter{})
+		for _, v := range tt.verifications {
+			c.now = func() time.Time { return t0.Add(v.after) }
 
-	c.ServeDNS(ctx, rec, req)
-	p = true // prefetch should be true for the 2nd fetch
-	c.ServeDNS(ctx, rec, req)
+			c.ServeDNS(context.TODO(), rec, req)
+			if v.fetch {
+				select {
+				case <-fetchc:
+					if !v.fetch {
+						t.Errorf("%s: want request to trigger a prefetch", tt.qname)
+					}
+				case <-time.After(time.Second):
+					t.Errorf("%s: want request to trigger a prefetch", tt.qname)
+				}
+			}
+			if rec.Rcode != dns.RcodeSuccess {
+				t.Errorf("%s: want rcode %d, got %d", tt.qname, dns.RcodeSuccess, rec.Rcode)
+			}
+			if want, got := 1, len(rec.Msg.Answer); want != got {
+				t.Errorf("%s: want %d answer RR, got %d", tt.qname, want, got)
+			}
+			if want, got := test.A(v.answer), rec.Msg.Answer[0]; !reflect.DeepEqual(want, got) {
```

Comment:
Not a superfan of using reflect here, can you just compare the strings?

Reply:
Interesting, I've never heard an argument against reflect in test usage, especially DeepEqual is super common in all go code bases. But as you wish, will change it.